### DISCUSSION TITLE
What are these non-fatal error logs I'm seeing; SQLiteLog: (1) no such column: raw_contact_id  ???

### DISCUSSION THREAD

**Discussion by reactormonk** [vestrel00/contacts-android#230] [Q&A] [answered]
My `logcat` is spitting out
```
06-22 16:03:34.600 10774 11967 E SQLiteLog: (1) no such column: raw_contact_id
06-22 16:03:34.601 10774 11967 E DatabaseUtils: Writing exception to parcel
06-22 16:03:34.601 10774 11967 E DatabaseUtils: android.database.sqlite.SQLiteException: no such column: raw_contact_id (code 1 SQLITE_ERROR): , while compiling: SELECT _id FROM view_contacts WHERE ((1)) AND (raw_contact_id IN ('35', '36', '37'))
06-22 16:03:34.601 10774 11967 E DatabaseUtils:         at android.database.sqlite.SQLiteConnection.nativePrepareStatement(Native Method)
06-22 16:03:34.601 10774 11967 E DatabaseUtils:         at android.database.sqlite.SQLiteConnection.acquirePreparedStatement(SQLiteConnection.java:903)
06-22 16:03:34.601 10774 11967 E DatabaseUtils:         at android.database.sqlite.SQLiteConnection.prepare(SQLiteConnection.java:514)
06-22 16:03:34.601 10774 11967 E DatabaseUtils:         at android.database.sqlite.SQLiteSession.prepare(SQLiteSession.java:588)
06-22 16:03:34.601 10774 11967 E DatabaseUtils:         at android.database.sqlite.SQLiteDatabase.validateSql(SQLiteDatabase.java:1795)
06-22 16:03:34.601 10774 11967 E DatabaseUtils:         at android.database.sqlite.SQLiteQueryBuilder.query(SQLiteQueryBuilder.java:530)
06-22 16:03:34.601 10774 11967 E DatabaseUtils:         at com.android.providers.contacts.ContactsProvider2.doQuery(ContactsProvider2.java:7207)
06-22 16:03:34.601 10774 11967 E DatabaseUtils:         at com.android.providers.contacts.ContactsProvider2.queryLocal(ContactsProvider2.java:7158)
06-22 16:03:34.601 10774 11967 E DatabaseUtils:         at com.android.providers.contacts.ContactsProvider2.queryDirectoryIfNecessary(ContactsProvider2.java:5575)
06-22 16:03:34.601 10774 11967 E DatabaseUtils:         at com.android.providers.contacts.ContactsProvider2.query(ContactsProvider2.java:5554)
06-22 16:03:34.601 10774 11967 E DatabaseUtils:         at android.content.ContentProvider.query(ContentProvider.java:1149)
06-22 16:03:34.601 10774 11967 E DatabaseUtils:         at android.content.ContentProvider$Transport.query(ContentProvider.java:241)
06-22 16:03:34.601 10774 11967 E DatabaseUtils:         at android.content.ContentProviderNative.onTransact(ContentProviderNative.java:102)
06-22 16:03:34.601 10774 11967 E DatabaseUtils:         at android.os.Binder.execTransact(Binder.java:731)

06-22 16:03:34.679 13903 13903 D ContactsDebug: Query.Result {
06-22 16:03:34.679 13903 13903 D ContactsDebug:     Number of contacts found: 3
06-22 16:03:34.679 13903 13903 D ContactsDebug:     First contact: Contact(id=37, lookupKey=0r35-2D3F29512F3929, rawContacts=[RawContact(id=35, contactId=37, addresses=[], emails=[], events=[], groupMemberships=[], ims=[], name=Name(id=205, rawContactId=35, contactId=37, isPrimary=false, isSuperPrimary=false, displayName=*******, givenName=*******, middleName=null, familyName=null, prefix=null, suffix=null, phoneticGivenName=null, phoneticMiddleName=null, phoneticFamilyName=null, isRedacted=true), nickname=null, note=null, organization=null, phones=[Phone(id=206, rawContactId=35, contactId=37, isPrimary=false, isSuperPrimary=false, type=null, label=null, number=****************, normalizedNumber=************, isRedacted=true)], photo=null, relations=[], sipAddress=null, websites=[], customDataEntities={}, isRedacted=true)], displayNamePrimary=*******, displayNameAlt=*******, lastUpdatedTimestamp=Wed Jun 22 16:03:34 GMT+02:00 2022, options=Options(id=37, starred=false, customRingtone=null, sendToVoicemail=false, isRedacted=true), photoFileId=null, photoUri=null, photoThumbnailUri=null, hasPhoneNumber=true, isRedacted=true)
06-22 16:03:34.679 13903 13903 D ContactsDebug:     isRedacted: true
06-22 16:03:34.679 13903 13903 D ContactsDebug: } (149 milliseconds)
```

But the code still does what's it's supposed to do :thinking:.

This is the code, it inserts a bunch of contacts, then adds pictures.

```
        val existingContacts = contacts.broadQuery().find()
        val deleteResult = contacts.delete().contacts(existingContacts).commit()
        log("existing contacts: $existingContacts - deleted all contacts $deleteResult")
        if (! deleteResult.isSuccessful && existingContacts.size > 0) { return false }

        val rawContacts = bundle.contacts.map {
            NewRawContact(
                name = NewName(
                    displayName = it.name
                ),
                phones = mutableListOf(
                    NewPhone(
                        number = it.number
                    )
                )
            )
        }

        val insertResult = contacts.insert().rawContacts(rawContacts).commit()
        log("inserted: $insertResult")
        if (! insertResult.isSuccessful) { return false }
        val insertedContacts = insertResult.contacts(contacts).zip(bundle.contacts)
        insertedContacts.forEach {
            val photoBitmap =
                BitmapFactory.decodeFile(bundleBase.resolve(it.second.picture).absolutePathString())
            it.first.setPhoto(contacts, photoBitmap)
        }
        log("Inserted Pictures")
```

**Comment by vestrel00** [comment]
@reactormonk, thanks for raising the issue, which I converted to a Q&A discussion. I also edited the title from,

"SQLiteLog: (1) no such column: raw_contact_id"

to,

"What are these non-fatal error logs I'm seeing; SQLiteLog: (1) no such column: raw_contact_id  ???"

I'll provide you the answer in my next comment.

**Comment by vestrel00** [comment, marked as answer]
This is expected behavior due to the way I've written a particular part of the `Query` API. Your code will run just fine with no side effects. There is no cause for concern except for the mystery of this unknown non-fatal error logs, which I am about to demystify 😁

## TLDR

It's an exception being suppressed internally by this library.

I can do something to reduce (maybe even eliminate) the occurrences of these error logs, so I have created an issue for it; #231

## What you can do if these error logs really bother you

If you want to get rid of these logs in your particular use case now, I suggest replacing this line of code;

```kotlin
val insertedContacts = insertResult.contacts(contacts).zip(bundle.contacts)
```

with;

```kotlin
val insertedContacts = contacts
    .query()
    .includeBlanks(false)
    .where { RawContact.Id `in` insertResult.rawContactIds }
    .find()
    .zip(bundle.contacts)
```

Setting `includeBlanks` to `false` (it's true by default) will prevent the error logs from occurring. Setting it to false should be completely fine for your particular chunk of code.

**OR**

You can wait until the next release, [0.2.3](https://github.com/vestrel00/contacts-android/discussions/233), which will contain the changes in #231

## More context

So, the code that causes this in the sample you provided is `insertResult.contacts(contacts)`. That is an extension function in [`InsertResult.kt`](https://github.com/vestrel00/contacts-android/blob/0.2.2/core/src/main/java/contacts/core/util/InsertResult.kt#L144), which invokes the following chunk of code;

```kotlin
contacts.query().where { RawContact.Id `in` rawContactIds }.find(cancel)
```

The `Query` API at some point in the `find()` function invocation will invoke the following [chunk of code](https://github.com/vestrel00/contacts-android/blob/main/core/src/main/java/contacts/core/Query.kt#L574-L581);

```kotlin
private fun ContentResolver.resolve(
    ...
    where: Where<AbstractDataField>?,
    ...
): List<Contact> {
    ...
    // Get Contact Ids matching where from the Data table. If where is null, skip.
    if (where != null ...) {
        ...
        // Get the Contacts Ids of blank RawContacts and blank Contacts matching the where from the
        // RawContacts and Contacts table respectively. Suppress DB exceptions because the where
        // clause may contain fields (columns) that are not in the respective tables.
        if (includeBlanks) {
            contactIds.addAll(
                findContactIdsInRawContactsTable(where.inRawContactsTable(), suppressDbExceptions  = true, cancel)
            )
            contactIds.addAll(
                findContactIdsInContactsTable(where.inContactsTable(), suppressDbExceptions = true, cancel)
            )
        }
        ...
    }
    ...
}
```

Notice that the `suppressDbExceptions` is set to true for this internal query to find blanks in the Contacts and RawContacts table. I'm intentionally suppressing the exceptions because the `where` that is provided by users of this library (or util functions) could contain columns that do not exist in the Contacts and RawContacts table.

Take this query for example;

```kotlin
fun getContactsWithAtLeastOneGmailEmail(): List<Contact> = Contacts(this)
    .query()
    .includeBlanks(true)
    .where { Email.Address endsWith "@gmail.com" }
    .find()
```

Email columns are only available in the Data table and do not exist in the Contacts and RawContacts table. This is guaranteed to cause an exception if used in a query in the Contacts and RawContacts table, which occurs when `includeBlanks` is set to true. Therefore, we suppress DB exceptions so we don't crash. I will not go into the purpose of `includeBlanks` here because that can be a whole other topic in of itself.

**Reply by vestrel00** [reply]
@reactormonk, please click the "Mark as answer" button if this answers your question 😁
